lsp: sort diagnostics by severity

mhinz opened the pull request:
Sort diagnostics (signs, virtual text) by severity, so that the most important message is shown first.

Fixes https://github.com/neovim/neovim/issues/13929

mhinz committed:
lsp: sort diagnostics by severity

Allow to sort diagnostics (and thus signs and virtual text) by severity, so that
the most important message is shown first.

    vim.lsp.handlers['textDocument/publishDiagnostics'] = vim.lsp.with(
      vim.lsp.diagnostic.on_publish_diagnostics, {
        severity_sort = true,
      }
    )

Fixes https://github.com/neovim/neovim/issues/13929

mjlbach commented on runtime/lua/vim/lsp/diagnostic.lua:
```diff
@@ -1086,6 +1087,11 @@ function M.display(diagnostics, bufnr, client_id, config)
     M.set_underline(diagnostics, bufnr, client_id, nil, underline_opts)
   end
 
+  local severity_sort = resolve_optional_value(config.severity_sort)
```
Should we maybe do this upstream as diagnostics come in rather than resorting on each display?

lithammer commented:
Maybe hold off on making it configurable until someone actually asks for it? Seems unlikely someone would want to turn sorting off.

mhinz commented:
Ok, we now sort uncondtionally just after the diagnostics arrive and before setting the diagnostics cache. ;)

mjlbach commented:
Should we have this be configurable? I'm not sure what the overhead is when we have several thousand diagnostics.

lithammer commented:
> Should we have this be configurable? I'm not sure what the overhead is when we have several thousand diagnostics.

Maybe we should put a configurable cap on the total number of diagnostics instead? I suspect having thousands of diagnostics is probably gonna be a poor experience for reasons other than slow sorting.

mjlbach commented:
> Maybe we should put a configurable limit on the total number of diagnostics instead

I think that's probably a separate PR. I for one, do not consider the order of diagnostics (and would turn it off), since this is probably the most expensive handler (since it's called unconditionally on every return from the language server) am a little concerned about adding any overhead. It's better than resorting on display though.

mhinz commented:
I understand the concerns.

We should just make `severity_sort` opt-in (and document the potential performance hit) and go on. :)

mfussenegger commented:
`get_line_diagnostics` already has a sort option and `set_loclist` also sorts by lnum.

Seems to me that this could be limited to virtual text if there is more than 1 diagnostic on a line to avoid duplicate sorting

mhinz commented:
@mfussenegger 

Currently the most important sign isn't necessarily shown first. That alone would justify such an option, IMHO.

And only sorting virtual text could lead to situations where you a see the sign for an "hint" but the virtual text for another "warning", which is even worse than the current situation. :)

mfussenegger commented:
@mhinz right I forgot about the signs. Still, multiple sorts don't seem ideal either. I think both of the other two sorts could be removed if the initial sort sorted by both lnum and severity, no?

neovim-discourse commented:
This pull request has been mentioned on **Neovim Discourse**. There might be relevant details there:

https://neovim.discourse.group/t/lsp-gutter-sign-priority/346/2

mhinz commented:
Okay, current status:

There are 3 uses of `table.sort`:

1. For `set_loclist()`. We don't care about that. It's only used on demand, so no issues with quickly sorting them again.
2. There was the option `sort_severity` for sorting diagnostics on `show_line_diagnostics()`. I removed that option. It always get sorted now, for the same reason as 1. It's only on demand, and the input is very small (I think there can only be 9 signs per line anyway?).
3. I introduced the `severity_sort` option for `vim.lsp.diagnostic.on_publish_diagnostics()`. It's the only `table.sort()` gets used regularly, so I made it opt-in.

mjlbach commented:
Since it's opt in I'm good with merging if you are :)

mjlbach commented:
Is there a compelling reason to remove the options though? Some people may use these

mhinz commented:
Then I'll merge it. For most (all) people nothing will change anyway, since I doubt that anyone used the _old_ `severity_sort = false` to get unsorted line diagnostics.. :)

mjlbach commented:
I thought previously if you didn't opt-in, diagnostics were unsorted for show_line_diagnostics? So this does change the default behavior

mhinz commented:
No, the _old_ `severity_sort` for `show_line_diagnostics()` was enabled by default, so opt-out.

mjlbach commented:
Ah my mistake, cool cool :)

folke commented:
@mhinz when there's multiple diagnostics for a line, we currently show the last, meaning with your change the one with the lowest severity.

https://github.com/neovim/neovim/blob/4de404a681426f9d5a1253ca7bc97b075ca14bee/runtime/lua/vim/lsp/diagnostic.lua#L784

That line should probably change to the first instead.

This way, if you have an error and an info diag on the same line, the error one will be displayed.

mhinz commented:
@folke 

Uhm, I can't reproduce that. We sort it by `>`, which means that the messages with the highest priority come last, which is what we want, because, with sign priority being the same for all signs, the last placed sign wins.

Therefore it seems correct that an error would be shown before an info diagnostic.

Or did I misundertand something?

folke commented:
@mhinz yeah, you're right. The problem happens when multiple clients set diagnostics. The sorting is per client. The last client overwrites the virtual text. 

In my case, I have client A finding an error on line 1 and B a warning. The warning is shown.

mjlbach commented:
Yeah multi-client support is another thing, it really wasn't anticipated people would be using multiple language servers per project, which seems to be popular with the javascript community, so a lot of these mechanisms weren't designed with this in mind... We're trying to make it better, but this could be a separate issue.

mhinz commented:
Oh.

Fixing that in a naive way would be easy, but then we would sort a tad too often. This needs some redesign in certan places.

Thanks for letting us know!

folke commented:
Makes sense. Thanks!

Tylerc230 commented:
For anyone wondering how to configure this setting: 
```vim.lsp.handlers["textDocument/publishDiagnostics"] = vim.lsp.with(
    vim.lsp.diagnostic.on_publish_diagnostics, {
        virtual_text = false,
        severity_sort = true
    }
)
```
Anywhere in your init.lua

spimanov commented:
In my case:
neovim-0.6.0~ubuntu1+git202109181149-1f49268c4-adeb5640f
nvim_lsp.pyright
nvim-lint
 I had to change the sort order:
```
vim.lsp.handlers["textDocument/publishDiagnostics"] = vim.lsp.with(
    vim.lsp.diagnostic.on_publish_diagnostics, {
        severity_sort = { reverse = true }
    }
)
```

gpanders commented:
@spimanov Should be fixed by https://github.com/neovim/neovim/pull/15701#event-5322875866
